### PR TITLE
Index configuration update

### PR DESCRIPTION
**Pull request by rdettai**
### Description

Add an index update endpoint. For this first version, only `search_settings` and `retetention_policy` can be configured.

Notes:
- the janitor has a mechanism to [update the index metadata every hour](https://github.com/quickwit-oss/quickwit/blob/24dc8d5eab5b700a362e5e31f3fcd58e8c395ee0/quickwit/quickwit-janitor/src/actors/retention_policy_executor.rs#L175-L188), so the retention policy update will be taken into account at that moment
- `list_indexes_metadata` calls [are not proxied through the control plane](https://github.com/quickwit-oss/quickwit/blob/24dc8d5eab5b700a362e5e31f3fcd58e8c395ee0/quickwit/quickwit-metastore/src/metastore/control_plane_metastore.rs#L126-L131), so [each `root_search` goes directly to the metastore](https://github.com/quickwit-oss/quickwit/blob/24dc8d5eab5b700a362e5e31f3fcd58e8c395ee0/quickwit/quickwit-search/src/root.rs#L1021-L1025) to get the `search_settings` at each request which means that these changes  are immediately visible by searches

### How was this PR tested?

This PR has a good unit test coverage. The fact that `search_settings` updates are immediately taken into account is covered by an integration test.

I also tested that the openapi spec is correctly rendered by swagger

### CODE REVIEW

File: quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs [hunk 1]
```diff
@@ -213,6 +213,11 @@ impl FileBackedIndex {
         &self.metadata
     }
 
+    /// Mutable ref to index config.
+    pub fn index_config_mut(&mut self) -> &mut IndexConfig {
+        &mut self.metadata.index_config
+    }
```
Comment (rdettai) [inline quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs:219]: if I understand correctly, it is okay to mutate any part of the `IndexMetadata`. But I wouldn't mind a second opinion from the reviewer.

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs [hunk 2]
```diff
@@ -399,6 +399,41 @@ impl MetastoreService for PostgresqlMetastore {
...
+        &mut self,
+        request: UpdateIndexRequest,
+    ) -> MetastoreResult<IndexMetadataResponse> {
+        let retention_policy_opt = request.deserialize_retention_policy()?;
+        let search_settings = request.deserialize_search_settings()?;
+        let index_uid: IndexUid = request.index_uid().clone();
+        let mut mutated_metadata_opt = None;
+        let mutated_metadata_ref = &mut mutated_metadata_opt;
+        run_with_tx!(self.connection_pool, tx, {
+            mutate_index_metadata::<MetastoreError, _>(
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:412]: Time to refactor `mutate_index_metadata` to return `MetastoreResult<MutationOccurred<T>, E>` instead of `MetastoreResult<bool, E>`.

File: quickwit/quickwit-metastore/src/tests/index.rs [hunk 3]
```diff
@@ -25,19 +25,24 @@
 //  - list_indexes
 //  - delete_index
 
+use std::vec;
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/tests/index.rs:28]: uh?

File: quickwit/quickwit-metastore/src/tests/index.rs [hunk 4]
```diff
@@ -78,6 +83,86 @@ pub async fn test_metastore_create_index<
...
+    let new_retention_policy_opt = Some(RetentionPolicy {
+        retention_period: String::from("3 days"),
+        evaluation_schedule: String::from("daily"),
+    });
+    assert_ne!(
+        index_metadata.index_config.retention_policy_opt, new_retention_policy_opt,
+        "original and updated value are the same, test became inefficient"
+    );
+
+    // run same update twice to check indempotence, then None as a corner case check
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/tests/index.rs:127]:
```suggestion
    // run same update twice to check idempotence, then None as a corner case check
```

File: quickwit/quickwit-metastore/src/tests/index.rs [hunk 5]
```diff
@@ -78,6 +83,86 @@ pub async fn test_metastore_create_index<
...
+        .index_metadata(IndexMetadataRequest::for_index_id(index_id.to_string()))
+        .await
+        .unwrap()
+        .deserialize_index_metadata()
+        .unwrap();
+
+    let new_search_setting = SearchSettings {
+        default_search_fields: vec!["body".to_string(), "owner".to_string()],
+    };
+    assert_ne!(
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/tests/index.rs:113]: That's overkill. Just pick a new field name that's out there or randomize it.

File: quickwit/quickwit-serve/src/index_api/rest_handler.rs [hunk 6]
```diff
@@ -521,6 +526,69 @@ async fn create_index(
         .await
 }
 
+/// The body of the index update request. All fields will be replaced in the
+/// existing configuration.
+#[derive(Deserialize, Serialize, Debug, Eq, PartialEq, Default, utoipa::ToSchema)]
+#[serde(deny_unknown_fields)] // Remove when adding new fields to allow to ensure forward compatibility
+pub struct IndexUpdates {
+    pub search_settings: SearchSettings,
+    /// The
```
Comment (guilload) [inline quickwit/quickwit-serve/src/index_api/rest_handler.rs:535]: The ? :)

File: docs/reference/rest-api.md [hunk 7]
```diff
@@ -301,8 +301,52 @@ curl -XPOST http://0.0.0.0:8080/api/v1/indexes --data @index_config.json -H "Con
...
+{
+    "search_settings": {
+        "default_search_fields": ["body"]
+    },
+    "retention": {
+        "period": "3 days",
+        "schedule": "@daily"
+    }
+}
+```
```
[the remark this answers is not in the document]
Comment (rdettai) [reply docs/reference/rest-api.md:342]: :+1:

File: quickwit/quickwit-metastore/src/tests/index.rs (hunk 3 above)
Comment (rdettai) [reply quickwit/quickwit-metastore/src/tests/index.rs:28]: oops!

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs [hunk 8]
```diff
@@ -288,26 +289,29 @@ macro_rules! run_with_tx {
     }};
 }
 
-async fn mutate_index_metadata<E, M: FnOnce(&mut IndexMetadata) -> Result<bool, E>>(
+async fn mutate_index_metadata<E, M>(
     tx: &mut Transaction<'_, Postgres>,
     index_uid: IndexUid,
     mutate_fn: M,
-) -> MetastoreResult<bool>
+) -> MetastoreResult<IndexMetadata>
```
Comment (rdettai) [inline quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:296]: - replaced mutable ref in mutation function with move to avoid a systematic clone
- didn't replace the returned mutation state (bool) as it wasn't used anywhere

File: quickwit/quickwit-metastore/src/metastore/index_metadata/mod.rs [hunk 9]
```diff
@@ -126,16 +126,16 @@ impl IndexMetadata {
         Ok(mutation_occurred)
     }
 
-    /// Deletes a source from the index. Returns whether the index was modified (true).
-    pub(crate) fn delete_source(&mut self, source_id: &str) -> MetastoreResult<bool> {
+    /// Deletes a source from the index.
+    pub(crate) fn delete_source(&mut self, source_id: &str) -> MetastoreResult<()> {
```
Comment (rdettai) [inline quickwit/quickwit-metastore/src/metastore/index_metadata/mod.rs:130]: I removed the returned mutation state flag as it forces callers to handle cases that never occur.

**Review by rdettai — COMMENTED** [review]
regarding the `mutate_index_metadata` refactoring commit:

File: quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs [hunk 10]
```diff
@@ -213,6 +213,20 @@ impl FileBackedIndex {
         &self.metadata
     }
 
+    /// Replace the search settings in the index config, returning whether a mutation occurred.
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs:216]:
```suggestion
    /// Replaces the search settings in the index config, returning whether a mutation occurred.
```

File: quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs [hunk 11]
```diff
@@ -213,6 +213,20 @@ impl FileBackedIndex {
...
+    pub fn set_search_settings(&mut self, search_settings: SearchSettings) -> bool {
+        let is_mutation = self.metadata.index_config.search_settings != search_settings;
+        self.metadata.index_config.search_settings = search_settings;
+        is_mutation
+    }
+
+    /// Replace the retention policy in the index config, returning whether a mutation occurred.
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/file_backed/file_backed_index/mod.rs:223]:
```suggestion
    /// Replaces the retention policy in the index config, returning whether a mutation occurred.
```

File: quickwit/quickwit-metastore/src/metastore/index_metadata/mod.rs (hunk 9 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/index_metadata/mod.rs:130]: Interesting. Usually, we err on the side of returning `Ok` if the target of the deletion does not exist or no longer exists. This makes those calls idempotent.

File: quickwit/quickwit-metastore/src/metastore/mod.rs [hunk 12]
```diff
@@ -179,6 +179,56 @@ impl CreateIndexResponseExt for CreateIndexResponse {
...
+pub trait UpdateIndexRequestExt {
+    /// Updates a new [`UpdateIndexRequest`] from an [`IndexConfig`].
+    fn try_from_updates(
+        index_uid: impl Into<IndexUid>,
+        search_settings: &SearchSettings,
+        retention_policy_opt: &Option<RetentionPolicy>,
+    ) -> MetastoreResult<UpdateIndexRequest>;
+
+    /// Deserializes the `index_config_json` field of a [`UpdateIndexRequest`] into an
+    /// [`IndexConfig`].
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/mod.rs:192]:
```suggestion
    /// Deserializes the `search_settings_json` field of an [`UpdateIndexRequest`] into a
    /// [`SearchSettings`] object.
```

File: quickwit/quickwit-metastore/src/metastore/mod.rs [hunk 13]
```diff
@@ -179,6 +179,56 @@ impl CreateIndexResponseExt for CreateIndexResponse {
...
+    fn deserialize_search_settings(&self) -> MetastoreResult<SearchSettings>;
+
+    /// Deserializes the `source_configs_json` field of a [`UpdateIndexRequest`] into an
+    /// `Vec` of [`SourceConfig`].
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/mod.rs:196]:
```suggestion
    /// Deserializes the `retention_policy_json` field of an [`UpdateIndexRequest`] into a
    /// [`RetentionPolicy`].
```

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs [hunk 14]
```diff
@@ -399,6 +403,34 @@ impl MetastoreService for PostgresqlMetastore {
...
+    ) -> MetastoreResult<IndexMetadataResponse> {
+        let retention_policy_opt = request.deserialize_retention_policy()?;
+        let search_settings = request.deserialize_search_settings()?;
+        let index_uid: IndexUid = request.index_uid().clone();
+        let updated_metadata = run_with_tx!(self.connection_pool, tx, {
+            mutate_index_metadata::<MetastoreError, _>(
+                tx,
+                index_uid,
+                |mut index_metadata: IndexMetadata| {
+                    if index_metadata.index_config.search_settings != search_settings
```
Comment (guilload) [inline quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:418]: Use the setters.

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs (hunk 8 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:296]: > * replaced mutable ref in mutation function with move to avoid a systematic clone

I don't see `.clone()` calls removed in this file. What am I missing?

> * didn't replace the returned mutation state (bool) as it wasn't used anywhere

👍

**Review by guilload — APPROVED** [review]

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs (hunk 8 above)
Comment (rdettai) [reply quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:296]: > I don't see .clone() calls removed in this file. What am I missing?

Actually you are right, the closure can completely use a mutable ref, it doesn't impact the ownership inside `mutate_index_metadata`. My bad!

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs (hunk 14 above)
Comment (rdettai) [reply quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs:418]: what setters are you referring to? The setters created in this PR are for the `FileBackedIndex` entity only, and I don't see any setters on `IndexMetadata`.

File: quickwit/quickwit-metastore/src/metastore/mod.rs (hunk 12 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/mod.rs]: No worries.

File: quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs (hunk 14 above)
Comment (guilload) [reply quickwit/quickwit-metastore/src/metastore/postgres/metastore.rs]: I thought the setters were `IndexMetadata`. My bad.

**Comment by dtherhtun** [comment]
what has happened to this PR. right now I couldn't update retention policy in existing index like this `quickwit index update retention-policy --index otel-traces-v0_7 --period "7 days" --schedule "daily"`
